ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/TestBlock.hs

```diff
@@ -0,0 +1,288 @@
+                      }
+
+treeToBlocks :: BlockTree -> [TestBlock]
+treeToBlocks = Tree.flatten . blockTree
+
+treeToChains :: BlockTree -> [Chain TestBlock]
+treeToChains = map (Chain.fromNewestFirst . reverse) . allPaths . blockTree
+
+treePreferredChain :: BlockTree -> Chain TestBlock
+treePreferredChain = fromMaybe Genesis
+                   . selectUnvalidatedChain
+                       testConfig
+                       Genesis
+                   . treeToChains
+
+instance Show BlockTree where
+  show (BlockTree t) = Tree.drawTree (fmap show t)
+
+instance Arbitrary BlockTree where
+  arbitrary = sized $ \n ->
+      BlockTree <$> mkTree 0.2 (map TestHash [1 .. max 1 (fromIntegral n)])
+  shrink (BlockTree t) =
+      BlockTree <$> shrinkTree t
+
+{-------------------------------------------------------------------------------
+  Generic auxiliary
+-------------------------------------------------------------------------------}
+
+-- | Construct random binary tree from given set of elements
+mkTree :: forall a.
+          Double -- ^ Likelyhood of branching at any point
+       -> [a] -> Gen (Tree a)
+mkTree threshold = go
+  where
+    go :: [a] -> Gen (Tree a)
+    go []     = error "go: no elements"
+    go [a]    = return $ Node a []
+    go (a:as) = do n <- choose (0, 1)
+                   if n >= threshold || null right
+                     then (\t   -> Node a [t])    <$> go as
+                     else (\l r -> Node a [l, r]) <$> go left <*> go right
+      where
+        (left, right) = split as
+
+-- | Shrink tree (without shrinking any elements)
+shrinkTree :: Tree a -> [Tree a]
+shrinkTree (Node a ts) = map (Node a) (shrinkList shrinkTree ts)
+                         -- Also try shrinking all subtrees at once
+                      ++ map (Node a) (transpose (map shrinkTree ts))
+
+
+-- | Split list into two
+--
+-- > split [1..6] == ([1,3,5],[2,4,6])
+-- > take 5 (fst (split [1..])) == [1,3,5,7,9]
+-- > take 5 (snd (split [1..])) == [2,4,6,8,10]
+split :: [a] -> ([a], [a])
+split []     = ([], [])
+split (a:as) = let (xs, ys) = split as in (a:ys, xs)
+
+-- | All paths through a tree
+allPaths :: Tree a -> [[a]]
+allPaths t = [] : nonEmptyPaths t
+
+nonEmptyPaths :: Tree a -> [[a]]
+nonEmptyPaths (Node a ts) = [a] : map (a:) (concatMap nonEmptyPaths ts)
+
+{-------------------------------------------------------------------------------
+  Test auxiliary
+-------------------------------------------------------------------------------}
+
+newtype Permutation = Permutation Int
+  deriving (Show)
+
+instance Arbitrary Permutation where
+  arbitrary = Permutation . cast <$> arbitrary
+    where
+      -- Use the generator for 'Int64' (rather than 'Int') as it is not biased
+      -- towards small values
+      cast :: Int64 -> Int
+      cast = fromIntegral
+
+  -- Doesn't make sense to shrink PRNG seed
+  shrink _ = []
+
+permutation :: Permutation -> [a] -> [a]
+permutation (Permutation n) = go (R.mkStdGen n)
```

Comment:
Nitpicking: wouldn't `permute` or `applyPermutation` be a better name?

Reply:
As you wish :)